acslib/ccure/crud.py

```diff
@@ -143,74 +101,100 @@ def add_image(
+            "Name": image_name,
+            "ParentId": personnel_id,
+            "ImageType": ImageType.PORTRAIT.value,
+            "PartitionID": partition_id,
+            "Primary": True,  # we're only adding primary portraits
+            "Image": image,
         }
-        return self.connection.request(
-            ACSRequestMethod.POST,
-            request_data=ACSRequestData(
-                url=self.config.base_url + self.config.endpoints.PERSIST_TO_CONTAINER,
-                data=self.connection.encode_data(request_data),
-                headers=self.connection.base_headers | self.connection.header_for_form_data,
-            ),
+        return self.add_children(
+            parent_type=ObjectType.PERSONNEL.complete,
+            parent_id=personnel_id,
+            child_type=ObjectType.IMAGE.complete,
+            child_configs=[image_properties],
         )
 
     def get_image(self, personnel_id: int) -> Optional[str]:
         """
         Get the `PrimaryPortrait` property for the person with the given personnel ID.
         The returned image is a base-64 encoded string.
         """
-        self.logger.info(f"Getting personnel image for personnel {personnel_id}")
-        request_json = {
-            "TypeFullName": "Personnel",
-            "pageSize": self.connection.config.page_size,
-            "pageNumber": 1,
-            "DisplayProperties": ["PrimaryPortrait"],
-            "WhereClause": f"ObjectID = {personnel_id}",
-        }
-        response = self.connection.request(
-            ACSRequestMethod.POST,
-            request_data=ACSRequestData(
-                url=self.config.base_url + self.config.endpoints.FIND_OBJS_W_CRITERIA,
-                request_json=request_json,
-                headers=self.connection.base_headers,
-            ),
-        ).json
-        if response:
-            return response[0]["PrimaryPortrait"]
-        raise ACSRequestException(
-            status.HTTP_400_BAD_REQUEST, f"No personnel found with ID {personnel_id}"
+        return self.get_property(self.type, personnel_id, "PrimaryPortrait")
+
+    def assign_clearances(self, personnel_id: int, clearance_ids: list[int]) -> ACSRequestResponse:
+        """Assign clearances to a person"""
+        clearance_assignment_properties = [
+            {"PersonnelID": personnel_id, "ClearanceID": clearance_id}
+            for clearance_id in clearance_ids
+        ]
+        return self.add_children(
+            parent_type=ObjectType.PERSONNEL.complete,
+            parent_id=personnel_id,
+            child_type=ObjectType.CLEARANCE_ASSIGNMENT.complete,
+            child_configs=clearance_assignment_properties,
+        )
+
+    def revoke_clearances(self, personnel_id: int, clearance_ids: list[int]) -> ACSRequestResponse:
+        """
+        Revoke a person's clearances
+        Two steps: 1: Get the PersonnelClearancePair object IDs
+                   2: Remove those PersonnelClearancePair objects
+        """
```

Comment:
Yeah, great question. I agree that we shouldn't be overfitting the CRUD implementations. However we can provide this/these action, but perhaps not bundled into the `crud` object itself.  Maybe a set of wrappers in a file `ccure/actions.py` that implements these multi step processes.

Reply:
It feels weird to do personnel CRUD actions with `ccure.personnel.search(...`, `ccure.personnel.create(...`, etc, but do other personnel actions with `ccure.actions.revoke_clearances(...`, `ccure.actions.get_personnel_image(...` etc. Especially since all of the non-CRUD methods we have defined here relate specifically to personnel.
Should we just rename `crud.py` to `actions.py` or something and not worry about sticking to CRUD?

Reply:
non-CRUD stuff is now moved to e.g `ccure.action.personnel.assign_clearances`